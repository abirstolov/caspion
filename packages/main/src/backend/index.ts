import { type Config } from '@/backend/commonTypes';
import { createTransactionsInExternalVendors } from '@/backend/export/exportTransactions';
import { scrapeFinancialAccountsAndFetchTransactions } from '@/backend/import/importTransactions';
import moment from 'moment';
import * as configManager from './configManager/configManager';
import * as Events from './eventEmitters/EventEmitter';
import outputVendors from './export/outputVendors';
import * as bankScraper from './import/bankScraper';
import logger from '../logging/logger';

export { CompanyTypes } from 'israeli-bank-scrapers-core';
export { Events, configManager, outputVendors };

export const { inputVendors } = bankScraper;

let intervalId: NodeJS.Timeout | null = null;

export async function scrapePeriodicallyIfNeeded(config: Config, optionalEventPublisher?: Events.EventPublisher) {
  const hoursInterval = config.scraping.periodicScrapingIntervalHours;
  optionalEventPublisher = optionalEventPublisher ?? new Events.BudgetTrackingEventEmitter();

  stopPeriodicScraping();

  if(hoursInterval && hoursInterval > 0) {
    await optionalEventPublisher.emit(Events.EventNames.LOG, { message: `Setting up periodic scraping every ${hoursInterval} minutes` });
    intervalId = setInterval(async () => {
      await scrapeAndUpdateOutputVendors(config, optionalEventPublisher);
    }, hoursInterval * 1000 * 60 * 60);
  }
}

export function stopPeriodicScraping() {
  if (intervalId) {
    clearInterval(intervalId);
  }
}

export async function scrapeAndUpdateOutputVendors(config: Config, optionalEventPublisher?: Events.EventPublisher) {
  const eventPublisher = optionalEventPublisher ?? new Events.BudgetTrackingEventEmitter();

  const startDate = moment().subtract(config.scraping.numDaysBack, 'days').startOf('day').toDate();

  await eventPublisher.emit(Events.EventNames.IMPORT_PROCESS_START, {
    message: `Starting to scrape from ${startDate} to today`,
  });

  const companyIdToTransactions = await scrapeFinancialAccountsAndFetchTransactions(
    config.scraping,
    startDate,
    eventPublisher,
  );
  try {
    const executionResult = await createTransactionsInExternalVendors(
      config.outputVendors,
      companyIdToTransactions,
      startDate,
      eventPublisher,
    );

    return executionResult;
  } catch (e) {
    logger.error('Failed to create transactions in external vendors', e);
    await eventPublisher.emit(
      Events.EventNames.GENERAL_ERROR,
      new Events.BudgetTrackingEvent({
        message: (e as Error).message,
        error: e as Error,
      }),
    );
    throw e;
  }
}
