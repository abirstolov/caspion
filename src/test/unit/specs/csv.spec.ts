import { TransactionStatuses, TransactionTypes } from 'israeli-bank-scrapers-core/lib/transactions';
import { EnrichedTransaction } from '@/backend/commonTypes';
import { serializeTransactions, parseTransactions } from '@/backend/export/outputVendors/csv/csv';

const transactions: EnrichedTransaction[] = [
  {
    identifier: 3982580,
    chargedAmount: 78,
    date: '2018-11-07T22:00:00.000Z',
    description: 'Ikea',
    installments: { number: 2, total: 7 },
    memo: 'hello',
    originalAmount: -50,
    originalCurrency: 'ILS',
    processedDate: '2018-11-10T22:00:00.000Z',
    accountNumber: 'abcd',
    category: 'Misc',
    hash: '111',
    status: TransactionStatuses.Completed,
    type: TransactionTypes.Normal
  },
  {
    identifier: undefined,
    chargedAmount: 1000,
    date: '2018-11-07T22:00:00.000Z',
    description: 'Samuel',
    installments: undefined,
    memo: 'amazing memo',
    originalAmount: 932,
    originalCurrency: 'ILS',
    processedDate: '2018-11-10T22:00:00.000Z',
    accountNumber: '8375982KJHDS2',
    category: 'Finance',
    hash: '11KJFLDKJ22__3231',
    status: TransactionStatuses.Pending,
    type: TransactionTypes.Installments
  }
];

describe('CSV exporter', () => {
  describe('Serialize and parse transactions', () => {
    test('Basic test case', () => {
      const serialized = serializeTransactions(transactions);
      expect(serialized).toMatchSnapshot();
      const parsed = parseTransactions(serialized);
      expect(parsed).toMatchSnapshot();
    });
    test('should convert undefined string to empty string', () => {
      transactions[1].memo = undefined;
      const serialized = serializeTransactions(transactions);
      const parsed = parseTransactions(serialized);
      expect(parsed[1].memo).toEqual('');
    });
  });
});
