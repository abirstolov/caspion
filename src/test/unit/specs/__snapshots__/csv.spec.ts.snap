// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSV exporter Serialize and parse transactions Basic test case 1`] = `
"date,description,chargedAmount,memo,category,accountNumber,installments,originalAmount,originalCurrency,processedDate,identifier,hash,status,type
2018-11-07T22:00:00.000Z,Ikea,78,hello,Misc,abcd,2 / 7,-50,ILS,2018-11-10T22:00:00.000Z,3982580,111,completed,normal
2018-11-07T22:00:00.000Z,Samuel,1000,amazing memo,Finance,8375982KJHDS2,,932,ILS,2018-11-10T22:00:00.000Z,,11KJFLDKJ22__3231,pending,installments
"
`;

exports[`CSV exporter Serialize and parse transactions Basic test case 2`] = `
Array [
  Object {
    "accountNumber": "abcd",
    "category": "Misc",
    "chargedAmount": 78,
    "date": "2018-11-07T22:00:00.000Z",
    "description": "Ikea",
    "hash": "111",
    "identifier": 3982580,
    "installments": Object {
      "number": 2,
      "total": 7,
    },
    "memo": "hello",
    "originalAmount": -50,
    "originalCurrency": "ILS",
    "processedDate": "2018-11-10T22:00:00.000Z",
    "status": "completed",
    "type": "normal",
  },
  Object {
    "accountNumber": "8375982KJHDS2",
    "category": "Finance",
    "chargedAmount": 1000,
    "date": "2018-11-07T22:00:00.000Z",
    "description": "Samuel",
    "hash": "11KJFLDKJ22__3231",
    "identifier": undefined,
    "installments": undefined,
    "memo": "amazing memo",
    "originalAmount": 932,
    "originalCurrency": "ILS",
    "processedDate": "2018-11-10T22:00:00.000Z",
    "status": "pending",
    "type": "installments",
  },
]
`;
